### DISCUSSION TITLE
Error from secp256k1 ECDSA `ecdsa_recover_internal`

### DISCUSSION THREAD

**Discussion by LawsonGraham** [aptos-labs/aptos-developer-discussions#450] [Questions] [answered]
### Describe your question in detail.

My goal is to securely confirm that a user's ETH address is owned by an Aptos address and vice versa - essentially a 'handshake'.

The way I want to do this is:

1. Have the ethereum wallet sign a message that is hash(aptos address)
2. Have the aptos account call a method `verify_eth_address` which is called with the user's aptos address signer, alongside the ECDSA signature created by the ethereum wallet

in `verify_eth_address`, we can verifyL:
 aptos address by asserting the hash(aptos address) == message signed
 eth address by recovering the signer public key, using keccak to hash that to get the eth address, and asserting recovered eth address == eth address passed in

I have implemented the logic, but I am getting the error:

native fun ecdsa_recover_internal(
│    │                ^^^^^^^^^^^^^^^^^^^^^^
│    │                │
│    │                Test was not expected to error, but it aborted with code 65537 originating in the module 0000000000000000000000000000000000000000000000000000000000000001::secp256k1 rooted here

I have confirmed my args are parse in correctly:

 let eth_address = x"<>"; // Your Ethereum address
        let message = x"<>"; // Your Aptos address as a message (32 bytes SHA-256 hash)
        let signature_bytes = x"efa2bc059fee02ae6ca419fe9b6b7e0825587a818290056729964bf1f81202ba74883695114c2dbf29fe2ed7cb06d3300e9eefaa0b6a19e6c4ab8e8ca39cb910"; // r + s combined
        let recovery_id = 27; // Recovery ID

### What error, if any, are you getting?

native fun ecdsa_recover_internal(
│    │                ^^^^^^^^^^^^^^^^^^^^^^
│    │                │
│    │                Test was not expected to error, but it aborted with code 65537 originating in the module 0000000000000000000000000000000000000000000000000000000000000001::secp256k1 rooted here

this `ecdsa_recover_internal` is a native function, so I can't see the logic to try and debug.

### What have you tried or looked at?  Or how can we reproduce the error?

I've verified args are parsed and loaded in correctly, the hashed messages line up too

I looked through aptos github for the error code, and nothing useful came up. Only relevant this is here: https://github.com/aptos-labs/nft-tutorial/blob/9d13e34eb5482f740f57b5c3e7db71e8d091e7e6/tutorial/step_5/BasicCoin/build/BasicCoin/sources/dependencies/AptosFramework/account.move#L643

this error suggests public key cannot be found... potential issue could be that ETH public keys aren't acceptable on aptos for some reason?

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

Aptos CLI

**Comment by LawsonGraham** [comment]
here is code framework btw:

```
public entry fun verify_accounts_entry(
        aptos_address: &signer,
        message: vector<u8>,
        recovery_id: u8,
        signature_bytes: vector<u8>,
        eth_address: vector<u8>
    ) {
        // Ensure the signature is 64 bytes (r + s)
        assert!(std::vector::length(&signature_bytes) == SIGNATURE_NUM_BYTES, ERR_MALFORMED_SIGNATURE);

        let signature = secp256k1::ecdsa_signature_from_bytes(signature_bytes);
        verify_accounts(aptos_address, message, recovery_id, &signature, eth_address);
    }

    public fun verify_accounts(
        aptos_address: &signer,  // The Aptos account (as the caller)
        message: vector<u8>,
        recovery_id: u8,         // The 'v' value from the Ethereum signature (recovery id)
        signature: &ECDSASignature,
        eth_address: vector<u8>      // The expected Ethereum address (20 bytes)
    ) {
        assert!(vector::length(&eth_address) == ETH_ADDRESS_SIZE, ERR_MALFORMED_ETH_ADDRESS);

        let hashed_message = sha2_256(message);

        let recovered_key_option = secp256k1::ecdsa_recover(hashed_message, recovery_id, signature);
        assert!(option::is_some(&recovered_key_option), ERR_SIGNATURE_RECOVERY_FAILED);
        let recovered_key = option::extract(&mut recovered_key_option);

        let recovered_eth_address = aptos_hash::keccak256(secp256k1::ecdsa_raw_public_key_to_bytes(&recovered_key));
        assert!(recovered_eth_address == eth_address, ERR_ETH_ADDRESS_MISMATCH);
        assert!(signer::address_of(aptos_address) == from_bcs::to_address(message), ERR_APTOS_ADDRESS_MISMATCH);
    }
    ```

**Comment by alinush** [comment]
Can you post the full code with the Move test and the input data please? 🙏

**Reply by LawsonGraham** [reply]
https://github.com/LawsonGraham/aptos-evm-signature/tree/main

**Reply by alinush** [reply]
In that code, `signature_bytes` is empty though (see [here](https://github.com/LawsonGraham/aptos-evm-signature/blob/c34b4af26f5baa45807dbfddd30035b4f21d9d64/aptos/signature_verifier/sources/eth_verifier.move#L91))...

**Reply by alinush** [reply]
I guess you are saying that running it with the following input still fails?

```
let eth_address = x"<>"; // Your Ethereum address
let message = x"<>"; // Your Aptos address as a message (32 bytes SHA-256 hash)
let signature_bytes = x"efa2bc059fee02ae6ca419fe9b6b7e0825587a818290056729964bf1f81202ba74883695114c2dbf29fe2ed7cb06d3300e9eefaa0b6a19e6c4ab8e8ca39cb910"; // r + s combined
let recovery_id = 27; // Recovery ID
```

**Reply by LawsonGraham** [reply]
just pushed some addresses, plz re take a look!

**Reply by alinush** [reply]
Oh, the problem is the `recovery_id`. I think it is wrong. It has to be `< 4`.

**Reply by alinush** [reply]
(Let me try and understand Ethereum's $r,s,v$ format. I think there is a discrepancy...)

**Reply by LawsonGraham** [reply]
@alinush looking at the ethers `signMessage` source code:
```
async signMessage(message: string | Uint8Array): Promise<string> {
        return this.signMessageSync(message);
    }
```

```
signMessageSync(message: string | Uint8Array): string {
        return this.signingKey.sign(hashMessage(message)).serialized;
    }
```
```
sign(digest: BytesLike): Signature {
        assertArgument(dataLength(digest) === 32, "invalid digest length", "digest", digest);

        const sig = secp256k1.sign(getBytesCopy(digest), getBytesCopy(this.#privateKey), {
            lowS: true
        });

        return Signature.from({
            r: toBeHex(sig.r, 32),
            s: toBeHex(sig.s, 32),
            v: (sig.recovery ? 0x1c: 0x1b)
        });
    }
```

this is the logic ethers takes to sign a message

I am not doing anything special with `v`... how do you recommend I resolve?

**Reply by alinush** [reply]
So apparently, the actual recovery ID is derived from `v`. 
 - for legacy transactions, the recovery ID is `v - 27`. This results in 0 or 1.
 - for EIP-155 transactions, the recovery ID is `(v - 35) % 2`. This also results in 0 or 1.
 
 I am not sure why recovery IDs 2 and 3 don't apply, but for now, try this ☝️?
 
 In your example above, replace `recovery_id = 27` with `recovery_id = 0`

**Reply by LawsonGraham** [reply]
ok - sounds good this resolved the error! thanks very much. One other thing I'm getting - for some reason I am getting a different public key derived from the ecdsa signature in the module logic vs. the public key which signed the message - any idea what's going wrong here?

**Reply by alinush** [reply]
Hm, not sure. May want to try the other recovery ID just in case you mistyped 28 as 27?

**Reply by LawsonGraham** [reply]
tried both with no luck - I'll look into it more, lmk if you've got any other ideas here. Afaik the message contents wouldn't impact the recovered public key right?

**Reply by alinush** [reply]
Oh, yes it will.

**Reply by alinush** [reply]
You need to input the right message, the right signature and the right recovery ID to get back the right PK.

**Reply by LawsonGraham** [reply]
@alinush I have debugged more, and found this:

ethers `wallet.signMessage` actually implicitly appends message data, causing the mismatch in hashes I was experiencing in my aptos module vs. in my ethers script.

I have since resolved this. So, I believe I have the correct message (both hashed messages passed into `sign` and `ecdsa_recover` respectively are the same) and the correct ID (27 => 0, 28 => 1). That leaves just the signature, where I am still getting the wrong public key back. I am passing in r  + s as the signature bytes... I believe this is correct? 

Could you take a look here? Given I don't have access to the source code and that this goes beyond my understanding of cryptography, I'm not totally sure how I could identify if there are differences in schemes or anything like this...

**Reply by alinush** [reply]
Make sure you are giving the signature in the correct (R,S) byte format? For example, big/little endianness may cause your issues.

This is the code that parses the signature from the 64 byte array: https://docs.rs/libsecp256k1/latest/src/libsecp256k1/lib.rs.html#532-545

If you are still not getting back the correct ETH address during recovery, it may be that you are not giving the correct signed message (maybe more hashing happens that you didn't notice).

Or, it may be that you are not correctly deriving the ETH address from the recovered pubkey? Again, endianness may get you. Or compressed vs. uncompressed PKs.

I think if you create some code that generates an ETH signature and some other code that verifies it in Move (which you already linked here), then I can further help you debug.

**Reply by alinush** [reply]
I think the code you posted above also does the signing, right?

**Reply by LawsonGraham** [reply]
@alinush yeah, everything I'm doing is included in the repo - feel free to try for yourself. I pushed info on what to do in the README.md. I stepped through all of the ethers signing logic, and I don't believe there's any extra hashing I'm not aware of in my current flow. I do `aptosAddress` -> `keccak256(aptosAddress)` -> signature on the message `keccak256(aptosAddress)` on ethers side.

As another confirmation of this, both the hash on aptos side and on eth side of the message are the same, so I believe it is in signature side of things

**Reply by alinush** [reply]
What about endianness? And (un)compressed PKs?

**Reply by LawsonGraham** [reply, marked as answer]
Ok I figured it out - it seems for some reason `SigningKey.computePublicKey(process.env.ETHEREUM_PRIVATE_KEY)` which I was using to view the public key given my PK has this property:

// raw public key; use uncompressed key with 0x04 prefix
        if (bytes.length === 64) {
            const pub = new Uint8Array(65);
            pub[0] = 0x04;
            pub.set(bytes, 1);
            bytes = pub;
        }

so they were the same public keys, it just appended 0x04 to the front (not sure why they do this)

So, it turns out that my signatures are producing the right output and recovering the Public Key correctly.
